### ISSUE TITLE
ExpireIn console sessions

### ISSUE THREAD

**Issue by waynebrantley** [pieceofsummer/Hangfire.Console#6]
Our jobs each have custom expire history settings.
So, I would like the ExpireIn to just go along with the job history setting.

Should I just set ExpireIn to super far in the future....and let it be removed with the job history (assume it will) or should there be another setting for expiration?

**Comment by pieceofsummer** [comment]
Unfortunately, expiring them along with a job is currently not possible.

Currently expiration happens in IServerFilter.OnPerformed, which knows nothing about job expiration. 

It would be known if we move it into IApplyStateFiler.OnStateApplied, but it can be executed on Dashboard side (where we lose track of console session itself), and even prior to actual job completion, so it becomes unreliable.

So I suggest to set ExpireIn to some reasonable time, like maximum expiration time for a job. For some jobs they will be kept longer than the job itself, but they still be eventually deleted.

**Comment by waynebrantley** [comment]
So, the console data is not tied to the deletion of the job?
Shouldn't it always be deleted when a job is...(or shortly after it)?

**Comment by pieceofsummer** [comment]
Generally speaking, it should, but it is not. 

The current Storage API doesn't support expiration of Set/Hash records along with a job. So this is a little trade-off of supporting all existing storage drivers.

**Comment by pieceofsummer** [comment]
More precisely, expiration of console sessions is counted from the end of a corresponding Processing state, i.e. if you have requeued the job multiple times, you will have several Processing states in history, each hosting its own console session, which will expire at different times.

**Comment by waynebrantley** [comment]
Ok.  To be clear:
The job (or requeue of job) will be completely removed by job retention policy,
However, the console session will remain in the system until it expires according to the config of your project?

Right?

Assuming that is right - I could potentially create a database specific job that looked for Set records without a job?  If so - I assume it is the Set table, how is that related to the job?

**Comment by pieceofsummer** [comment]
Yeah, you're right.

And yes, you can. Console sessions are stored in Set table (with some records in Hash, when the length doesn't fit Set table restrictions), you can see how console session identifier is serialized in ConsoleId.cs.

**Comment by waynebrantley** [comment]
So, it is something like this.
However, I cannot tell if it is one of your 'console logs' or not, so the substring is not that accurate, which is why I put the ExpireAt clause in there.  

  SELECT * FROM HangFire.[Set]
  WHERE ExpireAt IS NOT NULL AND  NOT EXISTS (SELECT * FROM HangFire.Job WHERE   id=SUBSTRING([Key],12,10))

If I could better recognize one of your entires, this would work better.  (like if it had something in the key that would let me determine it was from you)

I could look at the value I guess:
SELECT * FROM HangFire.[Set]
  WHERE LEFT(Value,5)='{"t":'
  AND NOT EXISTS (SELECT * FROM HangFire.Job WHERE id=SUBSTRING([Key],12,10))

Any other ideas?

**Comment by pieceofsummer** [comment]
I'd imagine a trigger like this:

```sql

create trigger TR_ExpireConsoleSessions 
on Hangfire.[Job]
after update
as
begin
	set nocount on;

	if update(ExpireAt)
	begin
		declare @ConsoleSessions table([Key] varchar(100) primary key, ExpireAt datetime not null);

		with ProcessingStates(JobId, StartedAt, ExpireAt) as (
			select s.JobId, 
				convert(datetime, convert(datetime2, substring(s.Data, charindex('"StartedAt":"', s.Data) + 13, 28))), 
				j.ExpireAt
			from inserted as j
				inner join Hangfire.[State] as s
					ON s.JobId = j.Id
			where j.ExpireAt is not null
			  and s.Name = 'Processing'
			  and charindex('"StartedAt":"', s.Data) > 0
		)
		insert @ConsoleSessions
		select lower(left(reverse(convert(varchar(100), convert(binary(6), 
					convert(bigint, datediff(second, '1970-01-01', StartedAt)) * 1000 + datepart(millisecond, StartedAt))
				, 2)), 11)) + convert(varchar(100), JobId) as [Key], 
				ExpireAt
		from ProcessingStates;

		-- expire Set records
		update s
		set ExpireAt = cs.ExpireAt
		from @ConsoleSessions as cs
			inner join Hangfire.[Set] as s
				on s.[Key] = cs.[Key]
		where s.ExpireAt <> cs.ExpireAt
		option (force order);

		-- expire Hash records
		update h
		set ExpireAt = cs.ExpireAt
		from @ConsoleSessions as cs
			inner join Hangfire.[Hash] as h
				on h.[Key] = cs.[Key]
		where h.ExpireAt <> cs.ExpireAt
		option (force order);
	end
end
```

**Comment by waynebrantley** [comment]
Different jobs need different amounts of history.  A job that runs once per week we might want to keep 6 months of history, where a job that runs once per day, we might want to keep only a month.

We have created a fully dynamic hangfire scheduler.   We have a job system where you specify what job runs and when through a web UI.  These jobs definitions are stored in a separate table.  We then have a custom IApplyStateFilter that looks at the JobHistoryRetentionInDays field of the job definition.  This gives us complete flexibility on setting it.

For this reason, there is no 'maximum possible job expiration'.  I mean 1 year is a 'reasonable maximum', but I would not really want to keep every console session for 1 year on a job that runs every 5 minutes, etc.   

So, I was hoping for a way to just clean them up.  My thought was if I set the console expire to one year...I would then be able to detect when the job was deleted and just clean it up automatically.   I could run another job that does it, run a sql agent job to do it, or add a trigger like you suggested above.

Any thoughts?

**Comment by pieceofsummer** [comment]
I might have an idea of how to expire console session along with a job without database-specific tricks, but it needs some tests to be done first.

Another approach may be to give users an explicit control over the session lifetime from the job itself, which overrides the global setting, something like:
```c#
context.SetConsoleExpireIn(TimeSpan.FromDays(1));
```

Either way, you need to use a workaround for now.

**Comment by waynebrantley** [comment]
Yes, SetConsoleExpireIn would be fine, because we could set it to the same dynamic days as the job.
Seems cleaner to not have any of these and have it automatically cleanup when the job is cleaned up.  :-)

**Comment by pieceofsummer** [comment]
I've finally implemented this in v1.3.0.

There's a new `FollowJobRetentionPolicy` option (enabled by default) to expire all consoles along with the parent job. May be disabled to use the old expiration algorithm based on `ExpireIn` value.

Please give it a try and see if it works as expected.
